diff --git a/drivers/spi/Makefile b/drivers/spi/Makefile
index b1bd4f9..23a22bc 100644
--- a/drivers/spi/Makefile
+++ b/drivers/spi/Makefile
@@ -10,6 +10,11 @@ ifeq ($(CONFIG_TRUSTONIC_TEE_SUPPORT),y)
 endif
 subdir-ccflags-y += -I$(srctree)/drivers/misc/mediatek/base/power/$(MTK_PLATFORM)
 ccflags-y += -I$(srctree)/drivers/misc/mediatek/include/mt-plat/
+
+ifeq ($(CONFIG_TRUSTKERNEL_TEE_SUPPORT),y)
+ccflags-y += -I$(srctree)/drivers/misc/mediatek/tkcore/include
+endif
+
 # small core, mostly translating board-specific
 # config declarations into driver model code
 obj-$(CONFIG_SPI_MASTER)		+= spi.o
diff --git a/drivers/spi/spi-mt65xx.c b/drivers/spi/spi-mt65xx.c
index 0af89cf..a07a684 100644
--- a/drivers/spi/spi-mt65xx.c
+++ b/drivers/spi/spi-mt65xx.c
@@ -27,6 +27,15 @@
 #include <linux/pm_runtime.h>
 #include <linux/spi/spi.h>
 
+#ifdef CONFIG_TRUSTKERNEL_TEE_SUPPORT
+#define SPI_TRUSTKERNEL_TEE_SUPPORT
+#endif
+
+#ifdef SPI_TRUSTKERNEL_TEE_SUPPORT
+#include <linux/tee_clkmgr.h>
+#include <linux/tee_fp.h>
+#endif
+
 #define SPI_CFG0_REG                      0x0000
 #define SPI_CFG1_REG                      0x0004
 #define SPI_TX_SRC_REG                    0x0008
@@ -752,6 +761,11 @@ static int mtk_spi_probe(struct platform_device *pdev)
 		goto err_disable_runtime_pm;
 	}
 
+#ifdef SPI_TRUSTKERNEL_TEE_SUPPORT
+    tee_clkmgr_register1("spi", ((1 << 15) - 2) - master->bus_num, 
+        clk_prepare_enable, clk_disable_unprepare, mdata->spi_clk);
+#endif
+
 	if (mdata->dev_comp->need_pad_sel) {
 		if (mdata->pad_num != master->num_chipselect) {
 			dev_err(&pdev->dev,
