commit a3fe9fa55c4f7cec55e8444a3d54561fb993b86c
Author: git <git@trustkernel.com>
Date:   Fri May 19 00:33:05 2017 +0800

    002.dts

diff --git a/arch/arm64/boot/dts/mt6735.dtsi b/arch/arm64/boot/dts/mt6735.dtsi
index e253d57..c4cec12 100755
--- a/arch/arm64/boot/dts/mt6735.dtsi
+++ b/arch/arm64/boot/dts/mt6735.dtsi
@@ -130,6 +130,11 @@ initrd=0x44000000,0x300000 loglevel=8 androidboot.hardware=mt6735";
 		<GIC_SPI 255 IRQ_TYPE_EDGE_RISING>;
 	};
 
+	tkcore {
+		compatible = "trustkernel,tkcore";
+		interrupts = <GIC_SPI 248 IRQ_TYPE_NONE>;
+	};
+
 	cpus {	#address-cells = <1>;
 		#size-cells = <0>;
 
